feat(remix-server-runtime): add `injectServerEffect`

kettanaito opened the pull request:
## Short description

This adds a new function called `injectServerEffect` that helps third-party packages establish process-wide server-side side effects in Remix. In particular, it implements effect persistence, allowing those side effects to survive HMR and HDR (since, by default, they won't and will inevitably get stale). 

This can be a great addition to Remix API for libraries like MSW, Prisma, and quite a few others. I'd like to use this in MSW examples since it yields a much simpler setup for the end developer, which is a win for everyone. I hope you consider merging this. 

## What's the intended usage of this?

```js
import { injectServerEffect } from '@remix-run/node'
import { init } from 'third-party'
import a from './a'

injectServerEffect('myEffect', (a) => {
  const customSideEffect = init(a)
  return () => {
    customSideEffect.anythingToCleanup()
  }
}, [a])
```

Here's an example of using the proposed API with MSW:

```js
// mocks/node.js
import { setupServer } from 'msw'
import { injectServerEffect } from '@remix-run/node'

export async function initMocks(...handlers) {
  await injectServerEffect('msw', (handlers) => {
    // This is the core of the effect: enabling request interception.
    const server = setupServer(...handlers)
    server.listen()

    // Return the instructions to cleanup.
    return () => {
      server.close()
    }
  }, [handlers])
}
```

<!--

👋 Hey, thanks for your interest in contributing to Remix!

If this is a simple docs change, go ahead and delete all this.

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a
lot of time and effort into a new feature. To avoid this from happening, we
request that contributors create a
[Feature Request discussion](https://github.com/remix-run/remix/discussions/new?category=ideas)
to first discuss any significant new features.

https://github.com/remix-run/remix/blob/main/CONTRIBUTING.md

Please fill in or delete each item below:

-->

Closes: https://github.com/remix-run/remix/discussions/6556

- [ ] Docs
- [x] Tests

Testing Strategy:

Testing this will be tricky since the feature is oriented at the custom effects surviving hot updates. I can at least try to provide this pull request with unit test and can count on the team's best judgement on how to properly test it in the integration scenarios.

<!--
Please explain how you tested this. For example:

> This test covers this code: <link to test>

Or

> I opened up my windows machine and ran this script:
>
> ```
> npx create-remix@0.0.0-experimental-7e420ee3 --template remix my-test
> cd my-test
> npm run dev
> ```
-->

kettanaito commented:
Testing this will be tricky since the feature is oriented at the custom effects surviving hot updates. I can at least try to provide this pull request with unit test and can count on the team's best judgement on how to properly test it in the integration scenarios.

kettanaito commented:
I'm also considering keeping all the effects in a global `Map` instead of letting Remix pollute the root-level global scope.

MichaelDeBoey commented:
@kettanaito Would you be willing to export it as in the Deno runtime as well + add tests please?

MichaelDeBoey commented:
@kettanaito Since this is a code change, could you rebase your branch onto latest `dev` & set target to `dev` as well?

kettanaito commented:
@MichaelDeBoey, I've rebased and set `dev` as the target. Also added some unit tests and exported the function from `/deno`. Happy to hear some feedback on this feature.

MichaelDeBoey commented on packages/remix-server-runtime/__tests__/injectServerEffect-test.ts:
```diff
@@ -0,0 +1,30 @@
+  getGlobalEffectById,
+} from "../injectServerEffect";
+
+describe(injectServerEffect, () => {
```
```suggestion
describe("injectServerEffect", () => {
```

kettanaito replied:
Can you elaborate more on why repeat the function name when you can infer it by passing the function reference to the `describe` block? 

The code I pushed yields `describe(injectServerEffect.name)`, which turns into your suggestion exactly.

MichaelDeBoey replied:
My suggestion is how we do it in the rest of the codebase, so I just want to keep it consistent

MichaelDeBoey commented on packages/remix-server-runtime/injectServerEffect.ts:
```diff
@@ -0,0 +1,133 @@
+  return effect;
+}
+
+export class ServerEffect<Dependencies extends Array<unknown>> {
```
Would it be possible to use normal objects + functions instead?
The only classes we currently use in the codebase are for creating custom `Error`s

kettanaito replied:
I don't mind. My motivation for using a class was to manage that self-updating `this.cleanup` reference. I can use the functional approach although, imo, it's a bit harder to understand than the class.

MichaelDeBoey replied:
Let's remove the class for consistency with the rest of the codebase (where we don't use classes)

herecydev commented on packages/remix-server-runtime/injectServerEffect.ts:
```diff
@@ -0,0 +1,133 @@
+export const SERVER_EFFECTS_KEY = Symbol("remixServerEffects");
+
+/**
+ * Creates a server-side effect persisted on tihs Node.js process.
```
```suggestion
 * Creates a server-side effect persisted on this Node.js process.
```

MichaelDeBoey replied:
```suggestion
 * Creates a server-side effect persisted on this Node.js process.
```

MichaelDeBoey reviewed (commented):
@kettanaito I tagged some people from the core Remix team so they can have a look at this and discuss further what direction they want to go with this

kettanaito commented:
I would really love to hear more opinions from the team before I dedicate more time to this feature. I've been talking with @pcattori, and it seems that Remix is not sure if they wish to add such a function officially.

kettanaito commented:
While this kind of API may still prove useful in some situations, speaking of MSW in particular, the current state of Remix allows to use initiate MSW server-side following regular Node.js integration rules. No extra steps are necessary anymore. I believe the credit goes to the recent reworks in HMR/HDR area by the Remix team. 

- [Remix + MSW example](https://github.com/mswjs/examples/tree/main/examples/with-remix)
